### DISCUSSION TITLE
How to inject viewmodel into page

### DISCUSSION THREAD

**Discussion by HakanL** [unoplatform/uno.extensions#1540] [Q&A] [answered]
I'm using UseNavigation to wire up my pages and viewmodels. I then navigate to the view model:
```
            await Navigator.NavigateViewModelAsync<HomeViewModel>(this);
```
and it works fine. But my question is/issue is that it will instantiate the HomePage first (without a DataContext), then the HomeViewModel, and then it will set the DataContext on the HomePage to the HomeViewModel. It all works, but I have some code in the HomePage that would be more efficient if the viewmodel was injected directly into the page code-behind. If I have a constructor that takes the VM then `new ViewMap<HomePage, HomeViewModel>()` complains.

**Comment by HakanL** [comment]
FWIW the `DataContext` on my `HomePage` is first assigned the `ShellViewModel` (the VM that navigates to the `HomePage`), then `null`, then the `HomeViewModel`. It seems inefficient/incorrect.

**Reply by nickrandolph** [reply]
The assignment of datacontext is a result of the view being attached to the visual tree and inheriting the datacontext from parent. From recollection we assign to null before the correct datacontext to ensure proper clean up of data bound elements.

**Reply by HakanL** [reply]
So it's not possible to inject the VM into the constructor of the page?

**Comment by nickrandolph** [comment, marked as answer]
Pages and Controls shouldn't take parameters in the constructors. In theory this rule could be broken for non-Page controls, since they could be instantiated from DI before being added to the visual tree. However, this will break hot reload scenarios in the future, so best to stick with the no-parameters in constructor rule.
The issue with Pages is that they get instantiated by a Frame as part of Frame.Navigate. This is somewhat painful as it means the navigation code doesn't get an opportunity to set the datacontext before the Page is navigated to

**Reply by HakanL** [reply]
Ok, I basically wanted to confirm what I see is the correct behavior, it sounds like it is so then I'm good.

**Reply by nickrandolph** [reply]
I think you're probably already doing this but the best way to trigger post-creation actions (that require access to the VM) is to hook the DataContextChanged event. When the DataContext is set to an instance of your viewmodel, you can perform the post-creation action.

Happy to take suggestions on lifecycle related matters (eg a callback on the view to initialize after the DataContext is set).

**Reply by HakanL** [reply]
Yes, that's exactly what I do, I was just surprised that it was first called with the VM from the previous page, then null, then my new VM. It may be correct, I was just surprised, and wanted to confirm that it was expected, and not something I've caused as I'm new to Uno, (but seasoned C# developer).
